File: ObjectiveC/FlyingLogo/Configuration.m
//
//  Configuration.m
//  FlyingLogo
//
//  Created by Gergely Sánta on 26/09/2019.
//  Copyright © 2019 TriKatz. All rights reserved.
//

#import "Configuration.h"
#import <ScreenSaver/ScreenSaver.h>

@interface Configuration() {
	NSUserDefaults *defaults;
}

@end

@implementation Configuration

+ (instancetype) shared
{
	static Configuration *configuration = nil;
	static dispatch_once_t onceToken;
	dispatch_once(&onceToken, ^{
		configuration = [[self alloc] init];
	});
	return configuration;
}

- (instancetype) init
{
	if (self = [super init]) {
		NSString *identifier = [NSBundle bundleForClass:Configuration.class].bundleIdentifier;
		defaults = [ScreenSaverDefaults defaultsForModuleWithName:identifier];

		_sceneNames = @[ @"Space Fly", @"Horizontal Flow" ];
		_scenes = @{
			@"Space Fly":       [NSNumber numberWithInt:SceneTypeSpaceFly],
			@"Horizontal Flow": [NSNumber numberWithInt:SceneTypeHorizontalFlow]
		};
	}
	return self;
}

- (SceneType) sceneType
{
	return (SceneType)[defaults integerForKey:@"sceneType"];
}

- (void) setSceneType:(SceneType)sceneType
{
	[defaults setInteger:sceneType forKey:@"sceneType"];
	[defaults synchronize];
}

@end

File: Swift/FlyingLogo/Configuration.swift
//
//  Configuration.swift
//  FlyingLogo
//
//  Created by Gergely Sánta on 28/09/2019.
//  Copyright © 2019 TriKatz. All rights reserved.
//

import Cocoa
import ScreenSaver

class Configuration {

	static let shared = Configuration()

	enum SceneType:Int {
		case spaceFly
		case horizontalFlow
	}

	let sceneNames = [ "SpaceFly", "HorizontalFlow" ]
	let scenes:[String:SceneType] = [
		"SpaceFly":       .spaceFly,
		"HorizontalFlow": .horizontalFlow
	]

	var sceneType:SceneType {
		get {
			return Configuration.SceneType(rawValue: defaults.integer(forKey: "sceneType")) ?? .spaceFly
		}
		set {
			defaults.set(newValue.rawValue, forKey: "sceneType")
			defaults.synchronize()
		}
	}

	private let defaults:UserDefaults

	private init() {
		if let identifier = Bundle(for: Configuration.self).bundleIdentifier {
			defaults = ScreenSaverDefaults(forModuleWithName: identifier) ?? ScreenSaverDefaults()
		} else {
			defaults = ScreenSaverDefaults()
		}
	}

}
